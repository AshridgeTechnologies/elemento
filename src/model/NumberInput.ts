import Element from './Element'
import {ElementId, ElementType, PropertyValueType} from './Types'
import BaseInputElement from './BaseInputElement'

export type Properties = {
    readonly initialValue?: PropertyValueType<number>,
    readonly label?: PropertyValueType<string>
}
export default class NumberInput extends BaseInputElement<Properties> implements Element {

    constructor(
        id: ElementId,
        name: string,
        properties: Properties) {
        super(id, name, properties)
    }

    static is(element: Element): element is NumberInput {
        return element.constructor.name === this.name
    }

    kind = 'NumberInput' as ElementType
}