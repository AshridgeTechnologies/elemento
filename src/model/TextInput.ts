import BaseElement from './BaseElement'
import Element from './Element'
import {ElementId, PropertyValue} from './Types'

export type Properties = {
    readonly initialValue?: PropertyValue,
    readonly maxLength?: PropertyValue,
    readonly multiline?: PropertyValue,
    readonly label?: PropertyValue
}
export default class TextInput extends BaseElement<Properties> implements Element {

    constructor(
        id: ElementId,
        name: string,
        properties: Properties) {
        super(id, name, properties)
    }

    static is(element: Element): element is TextInput {
        return element.constructor.name === this.name
    }

    static defaultValue = ''

    get initialValue() { return this.properties.initialValue }
    get maxLength() { return this.properties.maxLength }
    get multiline() { return this.properties.multiline }
    get label() { return this.properties.label }
}