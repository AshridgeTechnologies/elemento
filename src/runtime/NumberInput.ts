import React, {ChangeEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../util/helpers'
import {updateState, useObjectState} from './appData'

type Properties = {path: string, label?: string, initialValue?: number}

export default function NumberInput({path, initialValue = 0, label}: Properties) {
    const state = useObjectState(path)
    const optionalProps = definedPropertiesOf({label})
    const value = (state?.value !== undefined) ? state.value : initialValue
    const onChange = (event: ChangeEvent) => updateState(path, {value: Number((event.target as any).value) })

    return React.createElement(TextField, {
        id: path,
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value,
        onChange,
        ...optionalProps
    })
}
