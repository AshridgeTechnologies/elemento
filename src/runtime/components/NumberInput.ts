import React, {ChangeEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {PropVal, valueOfProps} from '../runtimeFunctions'
import InputComponentState from './InputComponentState'
import {useGetObjectState} from '../appData'

type Properties = {path: string, label?: PropVal<string>, }

export default function NumberInput({path, ...props}: Properties) {
    const {label} = valueOfProps(props)

    const optionalProps = definedPropertiesOf({label})
    const numericProps = {type: 'number', sx: { minWidth: 120, flex: 0 }}
    const state = useGetObjectState<NumberInputState>(path)
    const value = state._controlValue ?? ''
    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? Number(controlValue) : null
        state._setValue(updateValue)
    }

    return React.createElement(TextField, {
        id: path,
        variant: 'outlined',
        size: 'small',
        value,
        onChange,
        ...numericProps,
        ...optionalProps
    })
}

export class NumberInputState extends InputComponentState<number>  {
    defaultValue = 0
}

NumberInput.State = NumberInputState
