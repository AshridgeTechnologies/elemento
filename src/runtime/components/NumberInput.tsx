import React, {ChangeEvent, FocusEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {valueOfProps} from '../runtimeFunctions'
import InputComponentState from './InputComponentState'
import {NumberType} from '../types'
import {isNil, pick} from 'ramda'
import BigNumber from 'bignumber.js'
import DecimalType from '../types/DecimalType'
import BaseType from '../types/BaseType'
import {
    BaseInputComponentProperties,
    getLabelWithRequired,
    inputElementProps,
    propsForInputComponent,
    sxPropsForFormControl
} from './ComponentHelpers'
import {useObject} from '../appStateHooks'

type Properties = BaseInputComponentProperties

const decPlaces = (dataType: BaseType<any, any> | undefined) => dataType?.kind === 'Decimal' ? (dataType as DecimalType).decimalPlaces : undefined

const dataTypeProps = (dataType: BaseType<any, any> | undefined) => {
    const props = definedPropertiesOf(pick(['max', 'min'], dataType ?? {}))
    props.step = decPlaces(dataType) === 0 ? 1 : undefined
    return props
}

export default function NumberInput({path, ...props}: Properties) {
    const {label, readOnly, show, styles = {}} = valueOfProps(props)
    const sx = sxPropsForFormControl(styles, show, {minWidth: 120, flex: 0})

    const state: NumberInputState = useObject(path)
    const {dataValue, dataType} = state

    const formatValue = () => {
        if (isNil(dataValue)) return ''
        const dp = decPlaces(dataType)
        return dp !== undefined ? dataValue.toFixed(dp) : dataValue.toString()
    }
    const value = state.controlValue ?? formatValue()
    const labelWithRequired = getLabelWithRequired(dataType, label)
    const optionalProps = definedPropertiesOf({label: labelWithRequired})
    const inputComponentProps = propsForInputComponent(dataType, styles)
    const inputProps = inputElementProps(styles, readOnly, dataTypeProps(dataType))

    const error = state.errorsShown && !state.valid
    const helperText = state.errorsShown && state.errors ? (state.errors as string[]).join('.  ') : undefined

    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = () => {
            if (controlValue === '') return null
            if (dataType?.kind === 'Decimal') return new BigNumber(controlValue)
            return Number(controlValue)
        }
        state._setValues(updateValue(), controlValue)
    }
    const onBlur = (_event: FocusEvent) => state._setBlurred()

    return React.createElement(TextField, {
        id: path,
        variant: 'outlined',
        size: 'small',
        type: 'number',
        value,
        error,
        helperText,
        onChange,
        onBlur,
        sx,
        ...inputProps,
        ...inputComponentProps,
        ...optionalProps
    })
}

export class NumberInputState extends InputComponentState<number | BigNumber, NumberType | DecimalType>  {
    defaultValue = 0

    _setValues(value: number | BigNumber | null, editedValue: string) {
        this.updateState({value, editedValue})
    }
}

NumberInput.State = NumberInputState
