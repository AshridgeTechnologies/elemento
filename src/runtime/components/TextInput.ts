import React, {ChangeEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {updateState} from '../appData'
import {valueOfProps} from '../runtimeFunctions'

type Properties = {state: {value?: string, _path: string, _controlValue: string | null}, label?: string, maxLength?: number, width?: string | number, multiline?: boolean}

export default function TextInput({state, ...props}: Properties) {
    const {maxLength, label, multiline, width} = valueOfProps(props)
    const inputProps = maxLength !== undefined ? {inputProps: {maxLength}} : {}
    const widthProp = width !== undefined ? {width} : {}
    const sxProps = {sx: {...widthProp}}
    const optionalProps = definedPropertiesOf({label, multiline})
    const {_path: path} = state
    const value = state._controlValue ?? ''
    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? controlValue : null
        updateState(path, {value: updateValue })
    }

    return React.createElement(TextField, {
        id: path,
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value,
        onChange,
        ...inputProps,
        ...sxProps,
        ...optionalProps
    })
}
