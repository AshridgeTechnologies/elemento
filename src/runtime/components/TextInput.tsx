import React, {ChangeEvent, FocusEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {PropVal, valueOfProps} from '../runtimeFunctions'
import InputComponentState from './InputComponentState'
import {useGetObjectState} from '../appData'
import {TextType} from '../../shared/types'
import {pick} from 'ramda'
import {InputWithInfo} from './InputWithInfo'


type Properties = {path: string, label?: PropVal<string>, width?: PropVal<string | number>, multiline?: PropVal<boolean>, readOnly?: PropVal<boolean>}

export default function TextInput({path, ...props}: Properties) {
    const {label, multiline: multilineProp, width, readOnly} = valueOfProps(props)
    const widthProp = width !== undefined ? {width} : {}
    const sxProps = {sx: {...widthProp}}

    const state = useGetObjectState<TextInputState>(path)
    const {value} = state
    const dataType = state.dataType
    const required = dataType?.required
    const multiline = dataType?.format === 'multiline' || multilineProp
    const multilineProps = multiline ? {minRows: 2, maxRows: 10} : {}
    const optionalProps = definedPropertiesOf({label, multiline, ...multilineProps})

    const dataTypeProps = definedPropertiesOf(pick(['maxLength', 'minLength'], dataType ?? {}))
    const format = dataType?.format
    if (format && ['url', 'email'].includes(format)) {
        dataTypeProps.type = format
    }
    const inputPropsValues = {...dataTypeProps, ...definedPropertiesOf({readOnly}), }
    const inputProps = Object.keys(inputPropsValues).length > 0 ? {inputProps: inputPropsValues} : {}

    const error = state.errorsShown && !state.valid
    const helperText = state.errorsShown && state.errors ? (state.errors as string[]).join('.  ') : undefined

    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? controlValue : null
        state._setValue(updateValue)
    }
    const onBlur = (_event: FocusEvent) => state.ShowErrors(true)

    const formControl =  React.createElement(TextField, {
        id: path,
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value,
        error,
        helperText,
        onChange,
        onBlur,
        ...inputProps,
        ...sxProps,
        ...optionalProps
    })
    return InputWithInfo({description: dataType?.description, required, width, formControl})
}

export class TextInputState extends InputComponentState<string, TextType> {
    defaultValue = ''
}

TextInput.State = TextInputState