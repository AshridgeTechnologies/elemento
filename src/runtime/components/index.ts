export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
