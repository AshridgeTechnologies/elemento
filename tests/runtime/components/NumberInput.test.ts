/**
 * @jest-environment jsdom
 */

import {NumberInput, TextInput} from '../../../src/runtime/components/index'
import {snapshot, testAppInterface, wrappedTestElement} from '../../testutil/testHelpers'
import '@testing-library/jest-dom'
import {actWait, testContainer} from '../../testutil/rtlHelpers'
import {TextInputState} from '../../../src/runtime/components/TextInput'
import {NumberInputState} from '../../../src/runtime/components/NumberInput'
import {NumberType, TextType} from '../../../src/shared/types'
import {render} from '@testing-library/react'

const [numberInput, appStoreHook] = wrappedTestElement(NumberInput, NumberInputState)

const stateAt = (path: string) => appStoreHook.stateAt(path)

test('NumberInput element produces output with properties supplied',
    snapshot(numberInput('app.page1.width', {value: 27}, {label: 'Width'}))
)

test('NumberInput element produces output with default values where properties omitted',
    snapshot(numberInput('app.page1.height'))
)

test('NumberInput element produces output with description info', () => {
    const numberType = new NumberType('tt1', {description: 'A number input for entering numbers'})
    const {container} = render(numberInput('app.page1.description', {value: 27, dataType: numberType}, {label: 'Width'}))
    expect(container.innerHTML).toMatchSnapshot()
})

test('NumberInput shows value from the state supplied', () => {
    const {el} = testContainer(numberInput('app.page1.widget1', {value: 27}))
    expect(el`app.page1.widget1`.value).toBe('27')
})

test('NumberInput element produces output with properties supplied as state objects', () => {
    const {el} = testContainer(numberInput('app.page1.widget1', {value: 27},
        {label: new TextInputState({value: 'Item Number'})}))
    expect(el`label[for="app.page1.widget1"]`.innerHTML).toBe('Item Number')
})

test('NumberInput shows empty value when state value is absent', () => {
    const {el} = testContainer(numberInput('app.page1.widget1', {}))
    expect(el`app.page1.widget1`.value).toBe('')
})

test('NumberInput shows empty value when state value is set to undefined', () => {
    const {el} = testContainer(numberInput('app.page1.widget1', {value: undefined}))
    expect(el`app.page1.widget1`.value).toBe('')
})

test('NumberInput shows initial value when state value  exists', () => {
    const {el} = testContainer(numberInput('app.page1.widget1', {value: 99}))
    expect(el`app.page1.widget1`.value).toBe('99')
})

test('NumberInput shows empty value when state value is set to null and initial value exists', () => {
    const {el}  = testContainer(numberInput('app.page1.widget1', new TextInput.State({value: 'Axe'})._withStateForTest({value: null})))
    expect(el`app.page1.widget1`.value).toBe('')
})

test('NumberInput stores updated values in the app store section for its path', async () => {
    const {enter}  = testContainer(numberInput('app.page1.sprocket', {value: 27}))
    await enter('sprocket', '421')
    expect(stateAt('app.page1.sprocket').value).toBe(421)
})

test('NumberInput stores null value in the app store when cleared', async () => {
    const {el, user} = testContainer(numberInput('app.page1.sprocket', {value: 27}))
    const inputEl = el`app.page1.sprocket`
    await user.clear(inputEl)
    expect(stateAt('app.page1.sprocket')._controlValue).toBe(null)
})

test('NumberInput uses properties from dataType', async () => {
    const {el} = testContainer(numberInput('app.page1.sprocket2', {value: 27}, {label: 'The Input'}))

    const numberType = new NumberType('nt1', {min: 2, max: 10, format: 'currency'})
    await actWait(10)
    await actWait( () => stateAt('app.page1.sprocket2').props.dataType = numberType )

    await actWait( () => stateAt('app.page1.sprocket2').Reset() )
    expect(el`input`.min).toBe('2')
    expect(el`input`.max).toBe('10')
    expect(el`input`.type).toBe('number')
})

test('State class has correct properties', () => {
    const emptyState = new NumberInput.State({})
    expect(emptyState.value).toBe(0)
    expect(emptyState._controlValue).toBe(undefined)
    expect(emptyState.defaultValue).toBe(0)

    const state = new NumberInput.State({value: 77})
    const appInterface = testAppInterface(); state.init(appInterface, 'testPath')
    expect(state.value).toBe(77)
    expect(state.defaultValue).toBe(0)

    state.Reset()
    const resetState = state._withStateForTest({value: undefined, errorsShown: false})
    expect(appInterface.updateVersion).toHaveBeenCalledWith(resetState)
    expect(resetState.value).toBe(77)
    expect(resetState._controlValue).toBe(77)

    const clearedState = state._withStateForTest({value: null})
    expect(clearedState.value).toBe(0)
    expect(clearedState._controlValue).toBe(null)
})
