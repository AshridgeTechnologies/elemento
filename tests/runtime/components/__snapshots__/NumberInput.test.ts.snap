// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput element produces output with default values where properties omitted 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-yg1gz6-MuiFormControl-root-MuiTextField-root"
>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="app.page1.height"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="number"
      value=""
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-hdw1oc"
      >
        <span
          className="notranslate"
        >
          â€‹
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`NumberInput element produces output with properties supplied 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-yg1gz6-MuiFormControl-root-MuiTextField-root"
>
  <label
    className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    htmlFor="app.page1.width"
    id="app.page1.width-label"
  >
    Width
  </label>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="app.page1.width"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="number"
      value={27}
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-1in441m"
      >
        <span>
          Width
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
